### DISCUSSION TITLE
CosmosSerializationOptions to support other Newtonsoft options such as converters

### DISCUSSION THREAD

**Discussion by scottluskcis** [Azure/azure-cosmos-dotnet-v3#2064] [Q&A]
For the [CosmosJsonDotNetSerializer](https://github.com/Azure/azure-cosmos-dotnet-v3/blob/40ceaffa848ebb6e3f4efa255df49f38290ef539/Microsoft.Azure.Cosmos/src/Serializer/CosmosJsonDotNetSerializer.cs) it accepts a **CosmosSerializationOptions** object which then just gets mapped to **Newtonsoft JsonSerializerSettings**. This limits what we can do when want to do things such as using **Custom JSON Converters** because only the following properties are being used. 

```csharp
JsonSerializerSettings jsonSerializerSettings = new JsonSerializerSettings()
{
    NullValueHandling = cosmosSerializerOptions.IgnoreNullValues ? NullValueHandling.Ignore : NullValueHandling.Include,
    Formatting = cosmosSerializerOptions.Indented ? Formatting.Indented : Formatting.None,
    ContractResolver = cosmosSerializerOptions.PropertyNamingPolicy == CosmosPropertyNamingPolicy.CamelCase
        ? new CamelCasePropertyNamesContractResolver()
        : null
    };
```

Is there a reason why **CosmosSerializationOptions** doesn't support additional **JsonSerializerSettings**? Also the class is `internal` so prevents from any overrides of this behavior. 

Can the class either be made public and be extended or can there be a way to support more **JsonSerializerSettings**, that would be a great help. Opening discussion on this to get any inputs. Thanks for your feedback!

**Comment by scottluskcis** [comment]
Adding link to existing issue that is related [1813](https://github.com/Azure/azure-cosmos-dotnet-v3/issues/1813)

**Comment by j82w** [comment]
The reason is in the v4 SDK which will align with the [Azure Core SDK guidelines](https://github.com/Azure/azure-sdk-for-net/tree/master/sdk/core/Azure.Core) requires the SDK to not take a [dependency on Newtonsoft](https://azure.github.io/azure-sdk/dotnet_introduction.html#dotnet-dependencies), but instead use the new more performant [System.Text.Json](https://devblogs.microsoft.com/dotnet/try-the-new-system-text-json-apis/). To make it an easier conversion from the v3 to v4 SDK it was decided to not directly expose any dependencies to Newtonsoft.

**Reply by gpuchtel** [reply]
Bump. What's the status of removing the dependency on Newtonsoft? In v3? (since I heard v4 is abandoned).

**Reply by SOTOLLC** [reply]
Bump.

**Reply by Pilchie** [reply]
We can't remove the dependency on Newtonsoft within the v3 SDK - doing so would be a breaking change.  Instead, we have been working on an option to allow the use of System.Text.Json instead. See https://github.com/Azure/azure-cosmos-dotnet-v3/pull/4332.
